### ISSUE TITLE
How to contribute to this repo?

### ISSUE THREAD

**Issue by chn-lee-yumi** [OpenMediaVault-Plugin-Developers/openmediavault-kvm#3]
Hi, I am using this plugin and want to add some feature to it. Any instructions?

**Comment by ryecoaaron** [comment]
I would fork the repo to a personal repo and submit a pull request with your changes.  The current code is in the middle of fairly large change.  So, it might be a bit until the next version can be released.

What features are you looking to add?  I use this plugin myself.

**Comment by chn-lee-yumi** [comment]
> I would fork the repo to a personal repo and submit a pull request with your changes. The current code is in the middle of fairly large change. So, it might be a bit until the next version can be released.
> 
> What features are you looking to add? I use this plugin myself.

I want to add "VM template" function so I can create multi VMs base on one image.

There are also some bugs on the frontend. For example, "remove disk" shows an empty list that I must modify the xml to remove it. I don't know if I can fix it but I will try.

**Comment by ryecoaaron** [comment]
Does the clone VM not work for you?  That would be very similar to a VM template function.

The frontend issues like remove disk are the ones I am working on.

**Comment by chn-lee-yumi** [comment]
> Does the clone VM not work for you? That would be very similar to a VM template function.

That's useful. But this method will copy the full disk image, which will be slow if the disk is too big. That would be great if it supports linked clone. (Looks that libvirt doesn't support it now...)  The scene that will use linked clone is that if I need to test something in a cluster, I can create many VM quickly, and delete them after I have tested.

**Comment by ryecoaaron** [comment]
I am a VMware admin during the day.  So, I understand what you need.  It does look like kvm supports it - https://fabianlee.org/2018/09/24/kvm-implementing-linked-clones-with-a-backing-file/.  This might be tough to add to the current code but the new virt-install code I am writing to create new VMs, it might be easier.  I will look at this.

**Comment by ryecoaaron** [comment]
Seems like you just need to create a backing file and then you could select that file when creating new VMs.  I would guess cloning an existing VM using that backing file would do the same thing.

**Comment by chn-lee-yumi** [comment]
> 



> Seems like you just need to create a backing file and then you could select that file when creating new VMs. I would guess cloning an existing VM using that backing file would do the same thing.

That's a good idea! I will try it later.

By the way, I don't know if it is suitable to add an "enable ksm" function in this plugin.  KSM is useful when you create a lot VMs because it can merge same memory page to reduce memory usage. (It can be configured via `/sys/kernel/mm/ksm/`)

**Comment by ryecoaaron** [comment]
>  it is suitable to add an "enable ksm" function in this plugin

I will look to see how easy it would be to add.  The thing that worries me is any change that restart libvirtd because it restarts your VMs.

**Comment by ryecoaaron** [comment]
Here is the problem though (on an RPi4):
$ zgrep KSM /proc/config.gz 
# CONFIG_KSM is not set

**Comment by ryecoaaron** [comment]
> So it should check if the kernel supports KSM first

I agree.  Just one of the many things the code needs to do.  I will test but this should be easy to add.

**Comment by chn-lee-yumi** [comment]
Thanks a lot again!

**Comment by chn-lee-yumi** [comment]
> Have you tried using the https://packages.debian.org/bullseye/ksmtuned package? It makes sure it is enabled and does a bit more optimization.

Yes, I had. But I would like to tune it by myself. So I stop it after I read the KSM document.

**Comment by ryecoaaron** [comment]
I only asked because installing a package is easier than writing the rest of the code.  But I can just do a simple enable.

**Comment by chn-lee-yumi** [comment]
> I only asked because installing a package is easier than writing the rest of the code. But I can just do a simple enable.

Yes, it an easy way to enable KSM.

**Comment by ryecoaaron** [comment]
If I don't use the ksmtuned package, I have to write a service to do the same thing.  Is there a reason you can't customize in /etc/default/ksmtuned or /etc/ksmtuned.conf?

**Comment by chn-lee-yumi** [comment]
> If I don't use the ksmtuned package, I have to write a service to do the same thing. Is there a reason you can't customize in /etc/default/ksmtuned or /etc/ksmtuned.conf?

No, sorry I don't know it's configurable.

**Comment by ryecoaaron** [comment]
Please let me know if those config files allow your customizations.

**Comment by chn-lee-yumi** [comment]
`/etc/ksmtuned.conf` can fit my needs, thanks.

**Comment by ryecoaaron** [comment]
Great.  Thanks for the update.

**Comment by ryecoaaron** [comment]
I am still open to features you want to add or suggest by the way : )

**Comment by chn-lee-yumi** [comment]
> I would guess cloning an existing VM using that backing file would do the same thing.

Hi, I had tested the clone VM with a backing file, but it didn't work as we thought. It will merge the disk.

First I create a linked clone disk: `qemu-img create -f qcow2 -F qcow2 -b win10old.qcow2 linkedclone.qcow2`

Then I create a new VM and use this disk. Start the VM, it worked.

So I shutdown the VM, and try to clone it, using clone botton on the web.

It can be cloned, but the new disk file is as big as the backing disk.

I find the process through ps:
```
/usr/bin/qemu-img convert -f qcow2 -O qcow2 -o compat=1.1,lazy_refcounts srv/dev-disk-by-uuid-5b249b15-24f2-4796-a353-5ba789dc1e45/vm-hdd-raid5-pool/linkedclone.qcow2 srv/dev-disk-by-uuid-5b249b15-24f2-4796-a353-5ba789dc1e45/vm-hdd-raid5-pool/linkedclone-clone.qcow2
```

**Comment by ryecoaaron** [comment]
Ok.  The default virt-clone options are not doing the right thing for linked clones then.  I will probably have to use virt-clone with the --file option and manually clone the backing file to use with the --file option.  I haven't had a chance to test any of this though.

**Comment by ryecoaaron** [comment]
https://github.com/OpenMediaVault-Plugin-Developers/openmediavault-kvm/commit/d9d16958741dcb157629de87e29f2c4ca8822896
